compose: Add warning banner for search views.

josiah-tesfu opened the pull request:
Added a warning compose banner when composing a message in a search view that doesn't include all the messages in a conversation.

This is helpful in a situation where a user sends a message without the full context and forgets they are in a narrowed search view.

Fixes: #25893.

![issue-25893-rec](https://github.com/zulip/zulip/assets/71205057/d4a01fc9-2dab-44d3-bd3b-2f42af87cbac)


<details>
<summary>Self-review checklist</summary>


- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

amanagr commented:
@evykassirer would you be up to review this?

evykassirer commented:
@amanagr  I wrote most of the code in #26205, which was used to help open this PR, so I might not be the best candidate. But I can move it to integration review if that seems good enough!

amanagr commented on web/src/compose_actions.js:
```diff
@@ -70,6 +71,35 @@ function show_compose_box(msg_type, opts) {
+
+function show_compose_in_search_view_banner() {
+    // TODO: use compose_state.stream_id() when #25785 merges.
+    let stream_id = "";
+    const stream_name = compose_state.stream_name();
+    if (stream_name) {
```
We shouldn't show `Go to conversation` when `stream_name` is empty.

josiah-tesfu replied:
In that case, should we have a banner without a button which would only disappear once the narrow is exited?

amanagr replied:
Is there a reason we are doing this only for streams and not DMs in search results?

josiah-tesfu replied:
No, there isn't. I'll update to include DMs.

amanagr commented:
@josiah-tesfu thanks for working on this! Left a comment.

amanagr commented on web/src/compose_actions.js:
```diff
@@ -70,6 +71,38 @@ function show_compose_box(msg_type, opts) {
+        const search_view_banner = render_compose_banner({
+            banner_type: compose_banner.WARNING,
+            classname: compose_banner.CLASSNAMES.compose_in_search_view,
+            stream_id,
+            topic_name: compose_state.topic(),
```
There is no need to pass around `stream_id` and `topic_name` in this case since these value case be changed by the user
and then "Go to conversation" will take user to the wrong narrow. You can use `compose_state` to get stream / topic / DM name when user clicks on "Go to conversation".

josiah-tesfu replied:
That makes sense. However, my original commit didn't pass `stream_id` and `topic_name`. That change was inputted by @evykassirer, so I'm unsure whether I should amend it.

evykassirer replied:
Yeah sorry about that!

amanagr commented:
@josiah-tesfu thanks for the update! Left a few comments, waiting for your reply.

josiah-tesfu commented:
@amanagr I resolved the issues you brought up. The banner now works for dm's in addition to streams, and I removed unnecessary passing of stream data into the banner parameters.

amanagr commented on web/src/compose_actions.js:
```diff
@@ -70,6 +71,28 @@ function show_compose_box(msg_type, opts) {
     $(".new_message_textarea").css("min-height", "3em");
     compose_ui.set_focus(msg_type, opts);
+    const filter = narrow_state.filter();
+    if (filter && !filter.is_common_narrow()) {
```
You need `if (filter && !filter.is_search()) {` condition here. `!filter.is_common_narrow` will trigger it for most of the narrows.

josiah-tesfu replied:
The condition with `!filter.is_search()` doesn't work for triggering the banner, so I left it as `!filter.is_common_narrow()`.

amanagr replied:
Oh! sorry I meant `if (filter && filter.is_search()) {`. You shouldn't just use any condition that works.

josiah-tesfu replied:
Thanks, fixed!

amanagr commented:
@josiah-tesfu thanks for the update! Let's also squash both the commits into one.

josiah-tesfu commented:
@evykassirer are you familiar with using the `mock_template` for compose banners? I noticed you did something similar in a previous issue. Right now I'm failing the following test. I don't have much experience with zulip's testing so want to make sure I'm going in the right direction. I'm not sure what `data` would mean in this case, the general way tests should be written, and whether html or a zjquery stub is relevant here.

```
test failed: /__w/zulip/zulip/web/tests/compose_actions.test.js > start

Error: 
        Please use mock_template if your test needs to render compose_banner/compose_banner.hbs

        We use mock_template in our unit tests to verify that the
        JS code is calling the template with the proper data. And
        then we use the results of mock_template to supply the JS
        code with either the actual HTML from the template or some
        kind of zjquery stub.

        The basic pattern is this (grep for mock_template to see real
        world examples):

        run_test("test something calling template", ({mock_template}) => {
            // We encourage you to set the second argument to false
            // if you are not actually inspecting or using the results
            // of actually rendering the template.
            mock_template("compose_banner/compose_banner.hbs", false, (data) => {
                assert.deepEqual(data, {...};
                // or assert more specific things about the data
                return "stub-for-zjquery";
            });

            // If you need the actual HTML from the template, do
            // something like below instead. (We set the second argument
            // to true which tells mock_template that is should call
            // the actual template rendering function and pass in the
            // resulting html to us.
            mock_template("compose_banner/compose_banner.hbs", true, (data, html) => {
                assert.deepEqual(data, {...};
                assert.ok(html.startWith(...));
                return html;
            });
        });
```

evykassirer commented:
You can essentially copy the structure of other places this banner is stubbed. You're replacing the code that tries to display the banner with code that just checks that you're rendering it with the right data.

```
    mock_template("compose_banner/compose_banner.hbs", false, (data) => {
        assert.equal(data.classname, compose_banner.CLASSNAMES.topic_missing);
        assert.equal(
            data.banner_text,
            $t({defaultMessage: "Topics are required in this organization."}),
        );
        missing_topic_error_rendered = true;
    });
```

In this example, it checks that the banner is being rendered with the right classname and banner text, and also sets `missing_topic_error_rendered` to true so that we can test that the banner was "rendered". Let me know if this makes sense and if you have more questions!

josiah-tesfu commented:
@zulipbot add "integration review"

josiah-tesfu commented:
@zulipbot remove "maintainer review"

josiah-tesfu commented:
Ready for integration review!

amanagr commented:
I pushed a commit to fix the banner being displayed in empty search narrow sometimes. LGTM now.

timabbott commented:
Thanks for working on this @josiah-tesfu! I note a few bugs when playing with this:
* If you switch out of a search view in some way other than "go to conversation", the banner doesn't hide.
* I think `is_search` is the wrong check; we probably want to check `supports_collapsing_recipients` instead, which is whether this is a type of view where having a message means the view will also have replies to that message.
* If you open the compose box via say the `x` shortcut, and then switch to a stream recipient, it doesn't appear. I don't see a reason why stream messages would be any different from direct messages here -- the point is that you may not be seeing the latest messages in the place you're composing to even though you might think you are.

timabbott commented on web/src/compose_actions.js:
```diff
@@ -224,6 +224,10 @@ export function start(msg_type, opts) {
 
+    if (opts.trigger === "empty_narrow_compose") {
+        compose_recipient.set_selected_recipient_id("");
+    }
```
Hmm, I worry that this will incorrectly cause us to fail to reply to the current stream when your current view is an empty narrow for a topic; can you explain the bug you're fixing here in more detail?

amanagr replied:
It works fine since `compose_state.set_stream_id(opts.stream_id);` is being set later if there is a stream present.

The issue I am trying to fix is `recipient_id` being set in the current narrow for compose from previous narrow if the current narrow doesn't has a recipient like an empty search.

amanagr replied:
Pushed to add that as a comment.

timabbott commented:
@josiah-tesfu were you able to resolve the bugs that I noted above? I see you have pushed to this PR since then, but not comment explaining how you addressed the feedback.

josiah-tesfu commented:
> @josiah-tesfu were you able to resolve the bugs that I noted above? I see you have pushed to this PR since then, but not comment explaining how you addressed the feedback.

Thanks for the feedback! I just pushed a commit to resolve the comments you made. Now, if you switch out of a search view through the sidebar, the banner hides. Also, if you open the compose box via the `x` shortcut, the banner appears.

I'm not sure about using `supports_collapsing_recipients` over `is_search` though. From what I understood, the banner should appear when in a narrowed search view, which would be activated through entering a search term in the original narrow. This seems to match `is_search`. Let me know if there's something I'm not getting there, though.

timabbott commented on web/src/compose.js:
```diff
@@ -534,6 +534,31 @@ export function initialize() {
+
+    $("body").on("mouseup", ".topic-name", () => {
+        $(`.${CSS.escape(compose_banner.CLASSNAMES.compose_in_search_view)}`).remove();
+    });
```
What does his handler do? This `topic-name` selector is way too vague of a selector -- one could imagine all sorts of unrelated elements using that class name.

timabbott replied:
Oh, I see, you're trying to do something when the view changes -- yeah this is reliant on that particular click handler, but there's lots of other ways to change a view that should hide this banner.

I think you want to just clear this banner from `narrow.handle_middle_pane_transition` -- but to do that nicely, you'll want to define a `compose_banner.hide_compose_in_search_banner()` function to avoid internal compose component details/selectors in the `narrow.js` code.

timabbott commented:
> I'm not sure about using supports_collapsing_recipients over is_search though. From what I understood, the banner should appear when in a narrowed search view activated through entering a search term in the original narrow. This seems to match is_search. Let me know if there's something I'm not getting.

The point is that any view with `!supports_collapsing_recipients` is a form of search view -- just operator search rather than keyword search. And those views will all have this same failure mode -- that you don't see the whole conversation that you might be replying to.

timabbott commented on web/src/compose_actions.js:
```diff
@@ -70,6 +71,23 @@ function show_compose_box(msg_type, opts) {
+        classname: compose_banner.CLASSNAMES.compose_in_search_view,
+        banner_text: $t({
+            defaultMessage:
+                "You are composing a message from a search view, which may not include the latest messages in the conversation.",
```
maybe this should be more like "This view may not contain all messages in visible conversations."

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/pull/27864 for this issue.
